string names
string properties
string values
